src/psyclone/psyir/transformations/nemo_sign_trans.py

```diff
@@ -0,0 +1,175 @@
+        descendent of an assignment and will raise an exception if
+        this is not the case.
+
+        :param node: a SIGN BinaryOperation node.
+        :type node: :py:class:`psyclone.psyGen.BinaryOperation`
+        :param symbol_table: the symbol table.
+        :type symbol_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        :returns: 2-tuple of new schedule and memento of transform.
+        :rtype: (:py:class:`psyclone.nemo.NemoInvokeSchedule`, \
+                 :py:class:`psyclone.undoredo.Memento`)
+
+        '''
+        self.validate(node, symbol_table)
+
+        schedule = node.root
+        memento = Memento(schedule, self, [node, symbol_table])
+
+        oper_parent = node.parent
+        assignment = node.ancestor(Assignment)
+        # Create two temporary variables.
+        res_var = symbol_table.new_symbol_name("res_sign")
+        symbol_table.add(DataSymbol(res_var, DataType.REAL))
+        tmp_var = symbol_table.new_symbol_name("tmp_sign")
+        symbol_table.add(DataSymbol(tmp_var, DataType.REAL))
+
+        # Replace operator with a temporary (res_var).
+        oper_parent.children[node.position] = Reference(res_var,
+                                                        parent=oper_parent)
+
+        # res_var=ABS(B)
+        lhs = Reference(res_var)
+        rhs = UnaryOperation.create(UnaryOperation.Operator.ABS,
+                                    node.children[1])
+        new_assignment = Assignment.create(lhs, rhs)
+        new_assignment.parent = assignment.parent
+        assignment.parent.children.insert(assignment.position, new_assignment)
+
+        # Replace the ABS intrinsic with inline code.
+        abs_trans = NemoAbsTrans()
+        abs_trans.apply(rhs, symbol_table)
+
+        # tmp_var=A
+        lhs = Reference(tmp_var)
+        new_assignment = Assignment.create(lhs, node.children[0])
+        new_assignment.parent = assignment.parent
+        assignment.parent.children.insert(assignment.position, new_assignment)
+
+        # if condition: tmp_var<0.0
+        lhs = Reference(tmp_var)
+        rhs = Literal("0.0", DataType.REAL)
+        if_condition = BinaryOperation.create(BinaryOperation.Operator.LT,
+                                              lhs, rhs)
+
+        # then_body: res_var=res_var*-1.0
+        lhs = Reference(res_var)
+        lhs_child = Reference(res_var)
+        rhs_child = Literal("-1.0", DataType.REAL)
+        rhs = BinaryOperation.create(BinaryOperation.Operator.MUL,
+                                     lhs_child, rhs_child)
+        then_body = [Assignment.create(lhs, rhs)]
+
+        # if [if condition] then [then_body]
+        if_stmt = IfBlock.create(if_condition, then_body)
+        if_stmt.parent = assignment.parent
+        assignment.parent.children.insert(assignment.position, if_stmt)
```

Comment:
I've decided my earlier comment about the ordering of output code must be wrong as this generates the correct code (according to one of the tests below). I don't understand how consistently inserting new code at `assignment.position` doesn't reverse the order. Or does `assignment.position` dynamically update so that we're always inserting new code just before the assignment?

Reply:
Yes assignment.position will give the current position of Assignment so if the tree changes it updates appropriately. I think it does something along the lines of `self.parent.children.index(self)`